botnim/collect_sources.py

```diff
@@ -1,47 +1,320 @@
+                return extracted_data
+            except Exception as recovery_error:
+                logger.error(f"Recovery attempt failed: {str(recovery_error)}")
+                # Return a minimal valid structure instead of error
+                return {
+                    "DocumentMetadata": {
+                        "DocumentTitle": "Parsing Error",
+                        "Description": "Failed to parse API response"
+                    },
+                    "error": str(e),
+                    "raw_content": response.choices[0].message.content
+                }
+    except Exception as e:
+        logger.error(f"Error in extract_structured_content: {str(e)}")
+        return {"error": str(e)}
+
+def determine_document_type(file_path: Path) -> str:
+    """
+    Determine the document type by extracting it from the file name.
+    The convention is that the document type appears before the underscore.
+    
+    Args:
+        file_path (Path): Path to the source file
+        
+    Returns:
+        str: Document type, defaults to empty string if pattern not found
+    """
+    try:
+        filename = file_path.stem  # Get filename without extension
+        if '_' in filename:
+            doc_type = filename.split('_')[0]
+            return doc_type.strip()
+    except Exception as e:
+        logger.error(f"Warning: Could not determine document type from filename {file_path}: {e}")
+    
+    return ""  # Default type if pattern not found
+
 def collect_sources_files(config_dir, context_name, source):
     files = list(config_dir.glob(source))
     file_streams = [(f.name, f.open('rb'), 'text/markdown') for f in files]
     return file_streams
 
-def collect_sources_split(config_dir, context_name, source):
-    filename = config_dir / source
-    content = filename.read_text()
-    content = content.split('\n---\n')
-    file_streams = [io.BytesIO(c.strip().encode('utf-8')) for c in content]
-    file_streams = [(f'{context_name}_{i}.md', f, 'text/markdown') for i, f in enumerate(file_streams)]
-    return file_streams
+def collect_sources_split(config_dir, context_name, source, extract_metadata=False):
+    """
+    Collect sources from a split file with enhanced metadata extraction.
+    """
+    try:
+        filename = Path(config_dir) / source
+        logger.info(f"Processing split file: {filename}")
+        
+        content = filename.read_text()
+        sections = content.split('\n---\n')
+        logger.info(f"Split file into {len(sections)} sections")
+        
+        sources = []
+        for idx, section in enumerate(sections):
+            if section.strip():
+                file_name = f'{context_name}_{idx}.md'
+                section_bytes = section.strip().encode('utf-8')
+                mime_type = 'text/markdown'
+                
+                metadata = None
+                if extract_metadata:
```

Comment:
It seems that this large code portion below is more or less duplicated 4 times in this file - let's extract it into a single function?

Reply:
turin into `get_metadata_for_content`